characters_counter/characters_counter.py

```diff
@@ -1,16 +1,56 @@
-from collections import Counter
+import collections
+import re
+from collections import Counter, defaultdict
 from functools import cache
+from pathlib import Path
 
 REQUIREMENT_OCCURRENCE = 1
+NUM_OF_MOST_COMMON_CHARACTERS = 3
 
 
 @cache
 def count_characters_occurrence(string: str) -> int:
     """
-    The function returns the number of characters in the string occuring only
-    once.
+    The function returns the number of characters in the string occuring only once.
     """
     characters_counts = Counter(string)
     unique_occurrences = filter(lambda occurrence: (occurrence == REQUIREMENT_OCCURRENCE), characters_counts.values())
     unique_characters_number = sum(unique_occurrences)
     return unique_characters_number
+
+
+def get_frequent_characters_per_sentence(sentence: str) -> dict:
+    """
+    The helper function counts the frequency of each character in the input sentence, excluding spaces.
+    Single word is considered as a sentence as well.
+    The number of characters returned is controlled by the global constant "NUM_OF_MOST_COMMON_CHARACTERS".
+    :param sentence: str
+    :return: dict[str, int]
+    """
+    count_characters = Counter(sentence.replace(" ", ""))
+    return dict(count_characters.most_common(NUM_OF_MOST_COMMON_CHARACTERS))
+
+
+def count_character_frequency(data_input: Path) -> defaultdict[str, Counter[str]] | dict[str, int] | None:
```

Comment:
That is too broad type annotation

defaultdict[str, Counter[str]] | dict[str, int] | None

Looking on it, I would be confused what is actually coming from the function. Please optimize it

Reply:
Do you mean my function return different data: it can be defaultdict or dict or none at all? Hmmm... yes, you are right, it is a mess... I tried to be more determite in my code.
In new version my function returns only defaultdict[str, Counter[str]]. But for empty string it returns the same and I am not sure about is it right or not?